src/hotspot/share/utilities/bitMap.cpp

```diff
@@ -36,38 +36,67 @@ using idx_t = BitMap::idx_t;
 
 STATIC_ASSERT(sizeof(bm_word_t) == BytesPerWord); // "Implementation assumption."
 
+// For the BitMaps with allocators that don't support reallocate
 template <class BitMapWithAllocator>
-bm_word_t* GrowableBitMap<BitMapWithAllocator>::reallocate(bm_word_t* old_map, idx_t old_size_in_bits, idx_t new_size_in_bits, bool clear) {
-  size_t old_size_in_words = calc_size_in_words(old_size_in_bits);
-  size_t new_size_in_words = calc_size_in_words(new_size_in_bits);
+static bm_word_t* pseudo_reallocate(const BitMapWithAllocator& derived, bm_word_t* old_map, size_t old_size_in_words, size_t new_size_in_words) {
+  assert(new_size_in_words > 0, "precondition");
 
-  bm_word_t* map = NULL;
-  BitMapWithAllocator* derived = static_cast<BitMapWithAllocator*>(this);
+  bm_word_t* map = derived.allocate(new_size_in_words);
+  if (old_map != NULL) {
+    Copy::disjoint_words((HeapWord*)old_map, (HeapWord*) map,
+        MIN2(old_size_in_words, new_size_in_words));
+  }
 
-  if (new_size_in_words > 0) {
-    map = derived->allocate(new_size_in_words);
+  derived.free(old_map, old_size_in_words);
 
-    if (old_map != NULL) {
-      Copy::disjoint_words((HeapWord*)old_map, (HeapWord*) map,
-                           MIN2(old_size_in_words, new_size_in_words));
-    }
+  return map;
+}
```

Comment:
It could, though I see that function as an implementation detail of ArenaBitMap and ResourceBitMap, that I don't want to burden CHeapBitMap with. I also like if we can keep the definition of GrowableBitMap simple, so that it's easier to read and understand. So, I'd prefer to not have to do this. I hope that's OK.

Reply:
Fine with me.